File: src/ts/magma/api/io/Path.ts
import { IOError } from "../../../magma/api/io/IOError";
import { Option } from "../../../magma/api/option/Option";
import { Result } from "../../../magma/api/result/Result";
import { Iterable } from "../../../magma/api/collect/list/Iterable";
import { Iter } from "../../../magma/api/collect/Iter";
export interface Path {
	writeString(output: string): Option<IOError>;
	readString(): Result<string, IOError>;
	resolveSibling(siblingName: string): Path;
	walk(): Result<Iterable<Path>, IOError>;
	findFileName(): string;
	endsWith(suffix: string): boolean;
	relativize(source: Path): Path;
	getParent(): Path;
	query(): Iter<string>;
	resolveChildSegments(children: Iterable<string>): Path;
	resolveChild(name: string): Path;
	exists(): boolean;
	createDirectories(): Option<IOError>;
}

File: src/java/magma/api/io/Path.java
package magma.api.io;

import magma.api.collect.Iter;
import magma.api.collect.list.Iterable;
import magma.api.option.Option;
import magma.api.result.Result;

public interface Path {
    Option<IOError> writeString(String output);

    Result<String, IOError> readString();

    Path resolveSibling(String siblingName);

    Result<Iterable<Path>, IOError> walk();

    String findFileName();

    boolean endsWith(String suffix);

    Path relativize(Path source);

    Option<Path> findParent();

    Iter<String> query();

    Path resolveChildSegments(Iterable<String> children);

    Path resolveChild(String name);

    boolean exists();

    Option<IOError> createDirectories();
}
